envoy/config/xds_resources_delegate.h

```diff
@@ -0,0 +1,83 @@
+#pragma once
+
+#include "envoy/api/api.h"
+#include "envoy/common/optref.h"
+#include "envoy/common/pure.h"
+#include "envoy/config/subscription.h"
+#include "envoy/config/typed_config.h"
+#include "envoy/protobuf/message_validator.h"
+#include "envoy/service/discovery/v3/discovery.pb.h"
+
+namespace Envoy {
+namespace Config {
+
+/**
+ * An interface for hooking into xDS resource fetch and update events.
+ * Currently, this interface only supports the SotW (state-of-the-world) xDS protocol.
+ *
+ * Instances of this interface get invoked on the main Envoy thread. Thus, it is important for
+ * implementations of this interface to not execute any blocking operations on the same thread.
+ * Any blocking operations (e.g. flushing config to disk) should be handed off to a separate thread
+ * so we don't block the main thread.
+ */
+class XdsResourcesDelegate {
+public:
+  virtual ~XdsResourcesDelegate() = default;
+
+  /**
+   * Returns a list of xDS resources for the given authority and type. It is up to the
+   * implementation to determine what resources to supply, if any.
+   *
+   * This function is intended to only be called on xDS fetch startup, and allows the
+   * implementation to return a set of resources to be loaded and used by the Envoy instance
+   * (e.g. persisted resources in local storage).
+   *
+   * @param authority_id Unique id for the authority/control plane from which the resources are
+   *   obtained. Typically either a cluster name or a host name.
+   * @param resource_type_url The URL for the type of the resource (e.g. Secrets, Clusters, etc).
+   * @return A set of xDS resources for the given authority and type.
+   */
+  virtual std::vector<envoy::service::discovery::v3::Resource>
+  getResources(const std::string& authority_id, const std::string& resource_type_url) PURE;
+
+  /**
+   * Invoked when SotW xDS configuration updates have been received from an xDS authority, have been
+   * applied on the Envoy instance, and are about to be ACK'ed.
+   *
+   * @param authority_id Unique id for the authority/control plane from which the resources are
+   *   obtained. Typically either a cluster name or a host name.
+   * @param resource_type_url The URL for the type of the resource (e.g. Secrets, Clusters, etc).
+   * @param resources The resources sent by the authority for the given type.
+   */
+  virtual void onConfigUpdated(const std::string& authority_id,
```

Comment:
I think that `authority_id` should not be used here.
Eventually the authority ID will be part of the resource name. At the moment it seems you want to have a single place where a SotW message is confirmed to be the right one, so to me it is unclear when will the authority-id will be useful? Moreover, using the cluster/host name seems to not capture having multiple resources from a single source (that has the same cluster/host name).

Reply:
I've been thinking about it some more and I am not sure how we can do away with authority_id for SotW.  First as some background, I ran an integration test that delivers the same resource type and resource name from two different xDS sources and the Envoy processing code for xDS seemed to have no problem with that.

Ultimately, I think we need the authority (or some id for the ApiConfigSource, maybe a hash of the ApiConfigSource?) for two reasons:
(1) because of the possibility, however unlikely, that the same resource type and resource name could get served from two different xDS sources
(2) More importantly, because we have no other way to keep track of what the SotW is for a given resource type from a particular xDS source, so that we can overwrite it with a new list of resources when we receive an update.  This is the case for both named and wildcard subscriptions.  There's no way to know if we should append or remove and add the received resources on a DiscoveryResponse.  Even looping through the resource names to check each time we get a DiscoveryResponse may not be sufficient in the unlikely case of #1.

So I think we still need some notion of the authority/ApiConfigSource, but I think it would be better abstracted into a class, so instead of passing in authority_id + resource_type_url to the functions in this class, we take in an abstract `XdsSourceId`, of which there could be different implementations.

WDYT?